apollo-router/src/spec/schema.rs

```diff
@@ -362,6 +364,118 @@ impl std::ops::Deref for ApiSchema {
     }
 }
 
+/// A schema ID is the sha256 hash of the schema text.
+///
+/// That means that differences in whitespace and comments affect the hash, not only semantic
+/// differences in the schema.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Deserialize, Serialize)]
+pub(crate) struct SchemaHash(
+    /// The internal representation is a pointer to a string.
+    /// This is not ideal, it might be better eg. to just have a fixed-size byte array that can be
+    /// turned into a string as needed.
+    /// But `Arc<String>` is used in the public plugin interface and other places, so this is
+    /// essentially a backwards compatibility decision.
+    Arc<String>,
+);
+impl SchemaHash {
+    pub(crate) fn new(sdl: &str) -> Self {
+        let mut hasher = Sha256::new();
+        hasher.update(sdl);
+        let hash = format!("{:x}", hasher.finalize());
+        Self(Arc::new(hash))
+    }
+
+    /// Return the underlying data.
+    pub(crate) fn into_inner(self) -> Arc<String> {
+        self.0
+    }
+
+    /// Return the hash as a hexadecimal string slice.
+    pub(crate) fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+
+    /// Compute the hash for an executable document and operation name against this schema.
+    ///
+    /// See [QueryHash] for details of what's included.
+    pub(crate) fn operation_hash(
+        &self,
+        query_text: &str,
+        operation_name: Option<&str>,
+    ) -> QueryHash {
+        QueryHash::new(self, query_text, operation_name)
+    }
+}
+
+impl Display for SchemaHash {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0.as_str())
+    }
+}
+
+/// A query hash is a unique hash for an operation from an executable document against a particular
+/// schema.
+///
+/// For a document with two queries A and B, queries A and B will result in a different hash even
+/// if the document text is identical.
+/// If query A is then executed against two different versions of the schema, the hash will be
+/// different again, depending on the [SchemaHash].
+///
+/// A query hash can be obtained from a schema ID using [SchemaHash::operation_hash].
+// FIXME: rename to OperationHash since it include operation name?
+#[derive(Clone, Hash, PartialEq, Eq, Deserialize, Serialize)]
+pub(crate) struct QueryHash(
```

Comment:
This is not blocking but something we could address once we have all tests passing and we know for sure that everything is working.
It's a bit odd that we have the QueryHash declared inside the schema module. Maybe we should have a hash.rs module, which would also guarantee the privacy of the internal storage of the hash types.

Reply:
good idea